src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,116 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
 public class ArrayList<T> implements List<T> {
+    private static final int START_SIZE = 10;
+    private static final String OF_BOUND_MESSAGE = "Out of bound index in ArrayList";
+    private static final String OF_ABSENT_MESSAGE = "Element of ArrayList ain't present";
+    private T[] storage;
+    private int size;
+
+    public ArrayList() {
+        this.storage = (T[]) new Object[START_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        if (storage.length == size) {
+            increaseSize();
+        }
+        storage[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        inspectIndex(index, size + 1);
+        T[] tempStorage = (T[]) new Object[size - index];
+        System.arraycopy(storage, index, tempStorage, 0, tempStorage.length);
+        size = index;
+        add(value);
+        for (int i = 0; i < tempStorage.length; i++) {
+            add(tempStorage[i]);
+        }
```

Comment:
`System.arraycopy` called n+2 times.
Refactor your code to use it only twice.

Reply:
refactored